### ISSUE TITLE
Ankou's integration seems buggy

### ISSUE THREAD

**Issue by jonathanmetzman** [google/fuzzbench#392]
In https://www.fuzzbench.com/reports/2020-05-28/index.html I tested new the new preemptible strategy in an experiment comparing ankou to other fuzzers. However, ankou seems to fail a lot in ways that have nothing to do with the preemptible strategy. For example, in freetype2-2017, only 6 out of 20 trials of ankou that weren't terminated before 23 hours, actually produced corpus for 23 hours. When I took a look at the logs of one of the trials that lasted 23 hours (using this command `gsutil cat gs://fuzzbench-data/2020-05-28/experiment-folders/freetype2-2017-ankou/trial-147317/results/fuzzer-log.txt`) I see the following:
```
[run_fuzzer] Running target with Ankou
[run_fuzzer] Running command: ./Ankou -app /out/fuzz-target -i /out/seeds -o /out/corpus
2020/05/28 22:37:30 Invalid fuzz duration: time: invalid duration .
Persistent mode detected.
Deferred fork server detected.
len(inputs) = 2
len(seedPts) = 2
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/out/fuzzer.py", line 48, in fuzz
    subprocess.check_call(command)
  File "/usr/local/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['./Ankou', '-app', '/out/fuzz-target', '-i', '/out/seeds', '-o', '/out/corpus']' died with <Signals.SIGKILL: 9>.
```

**Comment by jonathanmetzman** [comment]
@zchcai

**Comment by jonathanmetzman** [comment]
Actually, I'm not sure this is the integration's fault. It might be Ankou's since we do have corpora for about 45 minutes of fuzzing. But @zchcai it would probably help get this fixed if you had a reproducer for this.

**Comment by zchcai** [comment]
Just check the trials running towards end (should be 20). I think the main issues are due to the trials for ankou with freetype2-2017 and harfbuzz-1.3.2.

  | ankou-82800
-- | --
bloaty_fuzz_target | 20
curl_curl_fuzzer_http | 19
freetype2-2017 | 6
harfbuzz-1.3.2 | 3
jsoncpp_jsoncpp_fuzzer | 19
lcms-2017-03-21 | 19
libpng-1.2.56 | 20
libjpeg-turbo-07-2017 | 20
libpcap_fuzz_both | 20
libxml2-v2.9.2 | 19
mbedtls_fuzz_dtlsclient | 19
openthread-2019-12-23 | 20
proj4-2017-08-14 | 17
re2-2014-12-09 | 19
systemd_fuzz-link-parser | 20
vorbis-2017-12-11 | 20
woff2-2016-05-06 | 20
zlib_zlib_uncompress_fuzzer | 17

**Comment by zchcai** [comment]
Hi @jonathanmetzman , could you provide the results for `gsutil cat gs://fuzzbench-data/2020-05-28/experiment-folders/bloaty_fuzz_target-ankou/trial-145945/results/fuzzer-log.txt`?

**Comment by jonathanmetzman** [comment]
http://commondatastorage.googleapis.com/fuzzbench-data/2020-05-28/experiment-folders/bloaty_fuzz_target-ankou/trial-145945/results/fuzzer-log.txt

**Comment by jonathanmetzman** [comment]
That trial seems to have finished.

**Comment by zchcai** [comment]
I ran a local test for these two benchmark, and some log info refers to the `initForkserver` implementation in Ankou:

https://github.com/SoftSec-KAIST/Ankou/blob/eecb60d7d29be62ac1ec41aa04f940bb5ea9e24d/system.go#L453-L459

```
$ cat experiment-folders/freetype2-2017-ankou/trial-2/results/fuzzer-log.txt
[run_fuzzer] Running target with Ankou
[run_fuzzer] Running command: ./Ankou -app /out/fuzz-target -i /out/seeds -o /out/corpus
2020/06/01 05:40:23 Invalid fuzz duration: time: invalid duration .
Persistent mode detected.
Deferred fork server detected.
2020/06/01 05:40:23 Fork server handshake failed: EOF
2020/06/01 05:40:23 Problem in starting frk srv
2020/06/01 05:40:23 Problem when writing in control pipe: invalid argument
2020/06/01 05:40:23 Error runnint seed: invalid argument (t=106.262µs).
2020/06/01 05:40:23 Problem when writing in control pipe: invalid argument
2020/06/01 05:40:23 Error runnint seed: invalid argument (t=18.362µs).
len(inputs) = 2
len(seedPts) = 0
Empty testCase.
2020/06/01 05:40:23 Problem when writing in control pipe: invalid argument
2020/06/01 05:40:23 No seed as input and default seed could not execute: invalid argument
2020/06/01 05:40:23 !!! Problem at initialization. Cannot fuzz :( !!!
```

The other benchmark is the same as the finished one above:

```
$ cat experiment-folders/bloaty_fuzz_target-ankou/trial-3/results/fuzzer-log.txt
[run_fuzzer] Running target with Ankou
[run_fuzzer] Running command: ./Ankou -app /out/fuzz_target -i /out/seeds -o /out/corpus
2020/06/01 05:40:23 Invalid fuzz duration: time: invalid duration .
Persistent mode detected.
Deferred fork server detected.
len(inputs) = 94
len(seedPts) = 94
```

**Comment by Jiliac** [comment]
I'm surprised the handshakes fails. Especially on only a few subjects. What would be an easy way for me to reproduce this locally?

**Comment by zchcai** [comment]
> I'm surprised the handshakes fails. Especially on only a few subjects. What would be an easy way for me to reproduce this locally?

Hi Valentin, right now I prepare an [initial implementation](https://github.com/zchcai/fuzzbench/tree/filesys) for Linux environment for a pure local testing. The running scripts shall be the following:

Before starting, please [install docker](https://docs.docker.com/engine/install/) and set it able to [run without sudo](https://docs.docker.com/engine/install/linux-postinstall/#manage-docker-as-a-non-root-user).

1. get the initial pure local support code
```
git clone -b filesys https://github.com/zchcai/fuzzbench.git && cd fuzzbench
```
2. prerequisites
```
git submodule update --init
sudo apt-get install -y build-essential python3-dev python3-venv
make install-dependencies
```
3. prepare local config file, replace `local_*` paths to two different absolute paths. Cloud-related configs (except `cloud_project`) are redundant but need to preserve now in this file.
```
cat experiment-config.yaml
trials: 1
max_total_time: 600
cloud_project: fuzzbench
cloud_compute_zone: us-central1-a
cloud_experiment_bucket: gs://shouldnotbeused
cloud_web_bucket: gs://shouldnotbeused
local_experiment_bucket: <absolute_path_to>/fuzzbench-data
local_web_bucket: <absolute_path_to>/fuzzbench-reports
cloud_sql_instance_connection_name: "fuzzbench:us-central1:postgres-experiment-db=tcp:5432"
local_experiment: true
gsutil_disabled: true
```
4. run with an different experiment name each time.
```
source .venv/bin/activate
export FORCE_LOCAL=1
PYTHONPATH=. python3 experiment/run_experiment.py --experiment-config experiment-config.yaml --benchmarks freetype2-2017 --experiment-name local_test_name --fuzzers ankou
```
5. Later, please check the log file:
```
vim <absolute_path_to>/fuzzbench-data/local_test_name/experiment-folders/freetype2-2017-ankou/trial-1/results/fuzzer-log.txt 
```

During the experiment is running, we can access the first layer docker by
```
docker exec -ti dispatcher-container /bin/bash
```
For the second layer docker, we can check the container's name to access to it.
```
docker ps
docker exec -ti <the related name> /bin/bash
```

**Comment by Jiliac** [comment]
I tried to run locally on freetype2-2017 and php_php-fuzz-parser with the `make debug-ankou-*` command and I can't reproduce this issue. Since you suspect resources like disk usage might be the issue, makes me think a limited RAM might also be a problem. In some situation, Ankou uses a lot of memory.

**Comment by inferno-chromium** [comment]
```
root@da1717a3c7a1:/out# free -mh
              total        used        free      shared  buff/cache   available
Mem:           187G         19G        4.1G        897M        164G        164G
Swap:          190G        118M        190G
```
Ankou definitely looks memory hungry and we dont have enough ram on bots (like 3.75 gb). this is from local run. @Jiliac , is there some way/settings to lower memory usage to work in our environment.

**Comment by Jiliac** [comment]
Ah yes, Ankou is definitely memory hungry. Thanks for finding the issue @inferno-chromium ! For simplicity and speed, I kept all the seeds, the PCA objects, and other elements in memory. Currently, there is no setting available to lower consumption. Which benchmark did you use, and after how long did you notice memory usage? I should I have time in the next several weeks to optimize this.

**Comment by inferno-chromium** [comment]
Even re2 benchmark was like 5 GB in a few minutes. 
Others like php one (also see list in https://github.com/google/fuzzbench/issues/392#issuecomment-636540280), it becomes like 9.8gb in a few minutes.

**Comment by Jiliac** [comment]
Ok, thanks! I'll see if I can reduce it to an acceptable level.

**Comment by inferno-chromium** [comment]
@Jiliac - friendly ping, it will be nice to have Ankou back in FuzzBench in the next month or two.

**Comment by Jiliac** [comment]
Hey @inferno-chromium ! I looked at this during the summer. The problem is due to the PCA, which triggers after about 30 minutes of fuzzing. This is the main technique Ankou uses from other fuzzers. The only solution I can think of right now would be to unload some of the content from memory on the disk. Which is equivalent to activating swap, and would anyway have a significant cost in performance.

**Comment by inferno-chromium** [comment]
> Hey @inferno-chromium ! I looked at this during the summer. The problem is due to the PCA, which triggers after about 30 minutes of fuzzing. This is the main technique Ankou uses from other fuzzers. The only solution I can think of right now would be to unload some of the content from memory on the disk. Which is equivalent to activating swap, and would anyway have a significant cost in performance.

In a real-world/prod scenario, fuzzing is usually spread across hundreds of low-end bots. So, it is good to keep memory limits in control. Should we close the issue then if you think this is infeasible to do ? We run for 23 hr, so how bad is the disk overhead on performance?

**Comment by Jiliac** [comment]
I can see how this makes things inconvenient for the common setup :/  We developed Ankou on large servers with many CPU cores, and having a single process instead of launching many individual ones. In this case, I think the economy of scale makes the memory cost per core become about the same.

When I experienced swap memory usage, it was basically the same as stopping the fuzzing campaign. The fuzzer became ineffective. It looks like it's going to be difficult to match Ankou to FuzzBench setup :( Thanks a ton for pushing its integration though!

**Comment by inferno-chromium** [comment]
> I can see how this makes things inconvenient for the common setup :/ We developed Ankou on large servers with many CPU cores, and having a single process instead of launching many individual ones. In this case, I think the economy of scale makes the memory cost per core become about the same.
> 
> When I experienced swap memory usage, it was basically the same as stopping the fuzzing campaign. The fuzzer became ineffective. It looks like it's going to be difficult to match Ankou to FuzzBench setup :( Thanks a ton for pushing its integration though!

What is the config you recommend. We can look at adding those kind of bots sometime ? Like Ideal/minimum config in cores, ram.

**Comment by Jiliac** [comment]
Thanks a lot for being so flexible! The memory consumed depends on the number of seeds which can change depending on the program under test. I didn't do any systematic performance benchmark, but I think something 8 cores (not hyper-threads) with 16Gb memory would be the best.
